File: src/tools/shell_exec.js
import { spawn } from 'child_process';
import { platform } from 'os';
import { tool } from './registry.js';

tool('shell_exec', 'Execute a shell command with full system privileges. Use for: running programs, installing packages, git, file management, system admin, compiling, process management, networking, etc.', {
  properties: {
    command: { type: 'string', description: 'Shell command to execute' },
    cwd: { type: 'string', description: 'Working directory (optional)' },
    timeout: { type: 'number', description: 'Timeout in seconds (default 120)' },
    env: { type: 'object', description: 'Extra environment variables (optional)' },
  },
  required: ['command'],
}, true, ({ command, cwd, timeout = 120, env }) => {
  return new Promise((resolve) => {
    const isWin = platform() === 'win32';
    const proc = spawn(command, {
      shell: isWin ? 'cmd.exe' : '/bin/sh',
      cwd: cwd || process.cwd(),
      timeout: timeout * 1000,
      env: env ? { ...process.env, ...env } : process.env,
      stdio: ['pipe', 'pipe', 'pipe'],
    });
    let stdout = '', stderr = '';
    proc.stdout.on('data', d => { stdout += d; });
    proc.stderr.on('data', d => { stderr += d; });
    proc.on('close', code => {
      let r = stdout;
      if (stderr) r += `\n[stderr]\n${stderr}`;
      if (code !== 0 && code != null) r += `\n[exit code: ${code}]`;
      resolve(r.slice(0, 50000) || '(no output)');
    });
    proc.on('error', e => resolve(`[error] ${e.message}`));
  });
});

File: starbot/tools/shell_exec.py
import subprocess
import platform
from . import tool

@tool(
    name="shell_exec",
    description="Execute a shell command and return output.",
    params={
        "properties": {
            "command": {"type": "string", "description": "Shell command to execute"}
        },
        "required": ["command"],
    },
    dangerous=True,
)
def shell_exec(command: str) -> str:
    try:
        shell = True
        if platform.system() == "Windows":
            args = ["cmd", "/c", command]
            shell = False
        else:
            args = command
        r = subprocess.run(
            args, shell=shell, capture_output=True, text=True, timeout=60
        )
        out = r.stdout
        if r.stderr:
            out += f"\n[stderr]\n{r.stderr}"
        if r.returncode != 0:
            out += f"\n[exit code: {r.returncode}]"
        return out or "(no output)"
    except subprocess.TimeoutExpired:
        return "[error] Command timed out (60s)"
    except Exception as e:
        return f"[error] {e}"
